modules/develop/pages/data-transforms.adoc

```diff
@@ -0,0 +1,254 @@
+
+== How data transforms work
+
+Within Repdanda, there's a single Raft controller that manages cluster information, including data transforms. On every shard (core), Redpanda knows what data transforms exist in the cluster, as well as metadata about the transform, such as input and output topics and environment variables. Data transforms take an input topic and map it to an output topic, running virtual machines (VMs) on the same shard as the input topic. 
+
+To run Wasm, Redpanda uses just-in-time (JIT) compilation, which compiles it in memory, writes it to executable space, then runs the directly translated machine code. When it becomes leader of a given partition on the input topic, Redpanda spins up Wasm VMs and fetches the function's executable code in memory. Redpanda writes the output, which may go to the same shard on the same machine, or to a different shard on the same machine, or to a different machine. Because all transform function metadata is in memory, Redpanda can stop and start transforms when partitions are moved. 
+
+image::shared:wasm_architecture.png[Wasm architecture in Redpanda]
+
+The transform function processes every record produced to an input topic and returns one or more records that are then produced to an output topic. Data transforms are applied to all partitions on an input topic. A record is processed after it has been successfully written to disk on the input topic. The record is then copied into the Wasm runtime, transformed, and copied back out, and then the broker writes the resulting record to the output topic. 
+
+In the following diagram, you see that Broker A has its own disk that it persistently writes to, and Broker B also has its own disk that it persistently writes to. A transform doesn't affect the original produced record: it happens in the background after the write finishes.
+
+image::shared:wasm_flow.png[Flow of record in Redpanda data transform]
+
+When you deploy a data transform, Redpanda pushes the generated `.wasm` files and metadata to the cluster, which then stores and replicates the bytecode and metadata using internal broker topics. Wasm is executed within a runtime in the broker. Each shard of Redpanda runs its own copy of the transform function. There is at most one runtime for a transform for any shard, and memory for each function must be reserved within the broker. This is achieved with the following cluster configuration properties: 
+
+- `wasm_per_core_memory_reservation`: Total amount of memory to reserve per shard 
+- `wasm_per_function_memory_limit`: Limit for an individual transform's memory
+
+There may only be `wasm_per_core_memory_reservation`/ `wasm_per_function_memory_limit` transforms deployed to a cluster at once. CPU time is dynamically allocated to the Wasm runtime and ensures that the code does not run forever and cannot block the broker from handling traffic or doing other work, such as Tiered Storage uploads.
+
+== Enable data transforms
```

Comment:
@rockwotj I've separated this into different pages, as you suggested. I do think it's easier to digest now. Please review the whole thing, since it looks & feels different now. Are the subheadings in "How Data Transforms Work" page OK? And should we provide any guidance about setting the properties?

Reply:
> Are the subheadings in "How Data Transforms Work" page OK? 

I am not sure if we generally use properties or configuration for the last one, otherwise they are wonderful!


> And should we provide any guidance about setting the properties?

Yes we should.

Basically you need to set them based on how many functions you have and how much memory you think a transform needs. For tinygo, the overhead is minimal (64KB) and you'll want enough memory for your input record, and output record to be in memory at the same time. For standard golang, the overhead is 16MB, in addition to the memory you need for your transform. I'm not sure if we're going to launch with standard go, but how does that sound?